### PR TITLE
Bug 1912907: Helm chart repository index can contain unresolvable relative URL's

### PR DESCRIPTION
**Pull request by bverschueren**
**Fixes:** 

https://bugzilla.redhat.com/show_bug.cgi?id=1912907

**Analysis / Root cause:**

The HelmChartRepositories resource proxies a repositories' index.yaml file as-is. Some Helm repositories (e.g. Nexus3) generate index files containing relative chart URL's. These relative URL's - without base uri - are unreachable while a Helm chart is instantiated from the Developer Console.

**Solution Description:**

Prepend the remote repositories' URL to charts' relative URL's, similar to how it is done in the native Helm client.

**Test setup:**

1. setup a nexus3:
~~~
$ helm repo add sonatype https://sonatype.github.io/helm3-charts/
$ helm install my-nexus-repository-manager sonatype/nexus-repository-manager --version 29.0.1 --set persistence.enabled=false --set nexus.securityContext=""
$ oc expose svc my-nexus-repository-manager
~~~ 
2. create helm repo:
~~~
$ curl -u admin:$(oc exec deployment/my-nexus-repository-manager -- cat /nexus-data/admin.password) -X POST "$(oc get route my-nexus-repository-manager  -o jsonpath='{.spec.host}')/service/rest/v1/repositories/helm/hosted" -H  "accept: application/json" -H  "Content-Type: application/json" -d "{  \"name\": \"helmtest\",  \"online\": true,  \"storage\": {    \"blobStoreName\": \"default\",    \"strictContentTypeValidation\": true,    \"writePolicy\": \"allow_once\"  },  \"cleanup\": {    \"policyNames\": [      \"string\"    ]  }}"
~~~
3. upload content:
~~~
$ curl https://charts.mirantis.com/charts/nginx-0.1.0.tgz -o nginx-0.1.0.tgz
$ curl -u admin:$(oc exec deployment/my-nexus-repository-manager -- cat /nexus-data/admin.password) "$(oc get route my-nexus-repository-manager  -o jsonpath='{.spec.host}')/repository/helmtest/" --upload-file nginx-0.1.0.tgz -v
~~~
4. create helmchartrepository:
~~~
$ cat <<EOF| oc apply -f -
apiVersion: helm.openshift.io/v1beta1
kind: HelmChartRepository
metadata:
  name: nexus
spec:
  connectionConfig:
    url: "http://$(oc get routes.route.openshift.io  my-nexus-repository-manager -o jsonpath='{.spec.host}')/repository/helmtest"
EOF
~~~
5. login to developer console and try to install the helm chart
[Developer Console] -> Add -> Helm Chart -> Nginx v0.1.0

### CODE REVIEW

File: pkg/helm/chartproxy/repos.go [hunk 1]
```diff
@@ -88,6 +88,17 @@ func (hr helmRepo) IndexFile() (*repo.IndexFile, error) {
...
+	for _, entry := range indexFile.Entries {
+		for i := len(entry) - 1; i >= 0; i-- {
+			for u := len(entry[i].URLs) - 1; u >= 0; u-- {
+				entry[i].URLs[u], err = repo.ResolveReferenceURL(hr.URL.String(), entry[i].URLs[u])
+				if err != nil {
+					klog.Errorf("Error resolving chart url for %v: %v", hr, err)
+				}
+			}
+		}
+	}
```
Comment (pedjak) [inline pkg/helm/chartproxy/repos.go:100]: Let's use ranges when possible:

```go
for _, chartVersions := range indexFile.Entries {
  for _, chartVersion := range chartVersions {
	  for i, url := range chartVersion.URLs {
		  chartVersion.URLs[i], err = repo.ResolveReferenceURL(hr.URL.String(), url)
		  if err != nil {
			  klog.Errorf("Error resolving chart url for %v: %v", hr, err)
		  }
	  }
  }
}
```

File: pkg/helm/chartproxy/repos_test.go [hunk 2]
```diff
@@ -262,11 +263,12 @@ func TestHelmRepoGetter_ListErrors(t *testing.T) {
...
-		name      string
-		url       string
-		httpCode  int
-		indexFile string
-		err       bool
+		name          string
+		url           string
+		httpCode      int
+		indexFile     string
+		resultDiffers bool
```
Comment (pedjak) [inline pkg/helm/chartproxy/repos_test.go:270]: so far we were assuming that `indexFile` is the expected one, but that might not be true anymore. Instead of introducing just a flag, and for future improvements, how about that we introduce `expectedIndexFile` field. If set, it is going to be used, otherwise, we continue using `indexFile` as backup in the test assertion part.

File: pkg/helm/chartproxy/repos_test.go [hunk 3]
```diff
@@ -335,9 +344,19 @@ func TestHelmRepo_IndexFile(t *testing.T) {
 				if err != nil {
 					t.Error(err)
 				}
-				if !reflect.DeepEqual(expectedIndex.Entries, index.Entries) {
+				if !tt.resultDiffers && !reflect.DeepEqual(expectedIndex.Entries, index.Entries) {
```
Comment (pedjak) [inline pkg/helm/chartproxy/repos_test.go:347]: see the comment above: we can revert this change, and create `expectedIndex` as following:

```go
var expectedIndex *repo.IndexFile
if (tt.expectedIndexFile != "") {
  expectedIndex, err = helmrepo.LoadIndexFile(tt.expectedIndexFile)
} else {
  expectedIndex, err = helmrepo.LoadIndexFile(tt.indexFile)
}
```

File: pkg/helm/chartproxy/repos_test.go [hunk 4]
```diff
@@ -335,9 +344,19 @@ func TestHelmRepo_IndexFile(t *testing.T) {
...
+				for _, entry := range index.Entries {
+					for i := len(entry) - 1; i >= 0; i-- {
+						for _, url := range entry[i].URLs {
+							_, err := neturl.Parse(url)
+							if err != nil {
+								t.Errorf("Index %v contains non-absolute URLs", index)
+							}
+						}
+					}
+				}
```
Comment (pedjak) [inline pkg/helm/chartproxy/repos_test.go:359]: after applying the above suggesting, this block is not needed anymore.

File: pkg/helm/chartproxy/testdata/sampleRepoIndex4.yaml [hunk 5]
```diff
@@ -0,0 +1,210 @@
+apiVersion: v1
```
Comment (pedjak) [inline pkg/helm/chartproxy/testdata/sampleRepoIndex4.yaml:1]: for better readability, can we please name this file to reflect the fact that the index contains relative urls, e.g. `sampleRepoIndexWithRelativeURLs.yaml`?

**Review by pedjak — CHANGES_REQUESTED** [review]
Thanks for contribution, please check the comments.

File: pkg/helm/chartproxy/repos_test.go (hunk 2 above)
Comment (bverschueren) [reply pkg/helm/chartproxy/repos_test.go:270]: While I agree using an `expectedIndexFile` would be the preferred way, I can't think of any sane way to compare both files. Since the `httptest.NewServer` randomly allocates an address:port pair that is assigned to `connectionConfig.url` and this commit prepends that url to the relative chart repository url's we can't reliably determine how the `expectedIndexFile` should look like. Any suggestions are welcome!
https://github.com/openshift/console/blob/ccd330f0fcb060cafcfab622654ca253f87b9bb8/pkg/helm/chartproxy/repos_test.go#L317-L317
https://github.com/openshift/console/blob/ccd330f0fcb060cafcfab622654ca253f87b9bb8/pkg/helm/chartproxy/repos_test.go#L329-L329
https://github.com/openshift/console/blob/ccd330f0fcb060cafcfab622654ca253f87b9bb8/pkg/helm/chartproxy/repos.go#L94

Comment (pedjak) [reply pkg/helm/chartproxy/repos_test.go:270]: Good point! Let's then drop usage http server. In order to do that we need to slightly refactory `httpRepo` struct:

```diff
diff --git a/pkg/helm/chartproxy/repos.go b/pkg/helm/chartproxy/repos.go
index 89e7548b6..5232447e1 100644
--- a/pkg/helm/chartproxy/repos.go
+++ b/pkg/helm/chartproxy/repos.go
@@ -36,25 +36,13 @@ const (
 )
 
 type helmRepo struct {
-	Name            string
-	URL             *url.URL
-	Disabled        bool
-	TLSClientConfig *tls.Config
+	Name       string
+	URL        *url.URL
+	Disabled   bool
+	httpClient func() (*http.Client, error)
 }
 
-type TLSConfigGetter interface {
-	Get() (*tls.Config, error)
-}
-
-func (repo helmRepo) Get() (*tls.Config, error) {
-	return repo.TLSClientConfig, nil
-}
-
-func (repo helmRepo) httpClient() (*http.Client, error) {
-	tlsConfig, err := repo.Get()
-	if err != nil {
-		return nil, err
-	}
+func httpClient(tlsConfig *tls.Config) (*http.Client, error) {
 	tr := &http.Transport{
 		TLSClientConfig: tlsConfig,
 	}
@@ -169,7 +157,7 @@ func (b helmRepoGetter) unmarshallConfig(repo unstructured.Unstructured) (*helmR
 			return nil, err
 		}
 	}
-	h.TLSClientConfig = crypto.SecureTLSConfig(&tls.Config{
+	tlsClientConfig := crypto.SecureTLSConfig(&tls.Config{
 		RootCAs: rootCAs,
 	})
 	if tlsReference != "" {
@@ -192,9 +180,12 @@ func (b helmRepoGetter) unmarshallConfig(repo unstructured.Unstructured) (*helmR
 			if err != nil {
 				return nil, err
 			}
-			h.TLSClientConfig.Certificates = []tls.Certificate{cert}
+			tlsClientConfig.Certificates = []tls.Certificate{cert}
 		}
 	}
+	h.httpClient = func() (*http.Client, error) {
+		return httpClient(tlsClientConfig)
+	}
 	return h, nil
 }
 
diff --git a/pkg/helm/chartproxy/repos_test.go b/pkg/helm/chartproxy/repos_test.go
index 80f1274d1..8b170a4dd 100644
--- a/pkg/helm/chartproxy/repos_test.go
+++ b/pkg/helm/chartproxy/repos_test.go
@@ -2,11 +2,10 @@ package chartproxy
 
 import (
 	"errors"
-	"fmt"
 	"io/ioutil"
 	"net/http"
-	"net/http/httptest"
 	neturl "net/url"
+	"os"
 	"reflect"
 	"testing"
 
@@ -26,6 +25,12 @@ type apiError struct {
 	msg      string
 }
 
+type RoundTripFunc func(req *http.Request) *http.Response
+
+func (f RoundTripFunc) RoundTrip(req *http.Request) (*http.Response, error) {
+	return f(req), nil
+}
+
 func onlyResult(obj interface{}, err error) interface{} {
 	return obj
 }
@@ -301,21 +306,10 @@ func TestHelmRepo_IndexFile(t *testing.T) {
 				Client: fake.K8sDynamicClient(),
 			}
 			url := tt.url
-
 			if url == "" {
-				ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-					w.Header().Set("Content-Type", "application/yaml")
-					w.WriteHeader(tt.httpCode)
-					if tt.indexFile != "" {
-						content, err := ioutil.ReadFile(tt.indexFile)
-						if err != nil {
-							t.Error(err)
-						}
-						fmt.Fprintln(w, string(content))
-					}
-				}))
-				url = ts.URL
+				url = "http://repo.foo.com"
 			}
+
 			repoCR := unstructured.Unstructured{
 				Object: map[string]interface{}{
 					"apiVersion": "helm.openshift.io/v1beta1",
@@ -335,7 +329,27 @@ func TestHelmRepo_IndexFile(t *testing.T) {
 			if err != nil {
 				t.Error(err)
 			}
+			if tt.url == "" {
+				repo.httpClient = func() (*http.Client, error) {
+					return &http.Client{
+						Transport: RoundTripFunc(func(req *http.Request) *http.Response {
+							resp := &http.Response{
+								StatusCode: tt.httpCode,
+							}
+							if tt.indexFile != "" {
+								r, err := os.Open(tt.indexFile)
+								if err != nil {
+									t.Error(err)
+								}
+								resp.Body = ioutil.NopCloser(r)
+							}
+							return resp
+						}),
+					}, nil
+				}
+			}
 			index, err := repo.IndexFile()
+
 			if tt.err && err == nil {
 				t.Errorf("Expected error %v but got %v", tt.err, err)
 			}

```

with the above, we can fully then mock http client and have fix repo url all the time.

Comment (bverschueren) [reply pkg/helm/chartproxy/repos_test.go]: Thanks for the very detailed suggestion!

**Review by pedjak — CHANGES_REQUESTED** [review]
Please squash commits and craft more precise commits message. `HelmChartRepository` does not refer relative URLs, it is `index.yaml` of a repo that might contain relative chart urls, and this fix addresses that.

**Comment by andrewballantyne** [comment]
@pedjak Please re-review

We have Today/Tomorrow (Depending on your timezone) and Friday to see this merged for 4.7. Please let me know if there are any concerns on getting this merged.

cc @rohitkrai03

**Comment by pedjak** [comment]
> #7711 failed to apply on top of branch "release-4.6":

@bverschueren many thanks for you contribution - would you mind to prepare similar PR against `release-4.6` branch? Automating cherry picking did not work.
